// contadorrest/contadorrest.js
var rest = require('restler');
var netConfig = require('netconfig');
var host = netConfig.getHost();
var port = netConfig.getPort();

module.exports = {

     newContador: function(token, contador, callback){
          contador.prefixo = encodeURIComponent(contador.prefixo);
          contador.descricao = encodeURIComponent(contador.descricao);
          if (contador.descricao != "" && contador.prefixo != ""){
                   rest.post('http://'+host+":"+port+'/notas/contador/new',
                   {
                       accessToken: token,
                       data: contador
                   })
                   .on('success', function(data,response){
                       callback(data);
                   })
                   .on('error', function(err, response){
                       callback({message:"Erro ao inserir Contador!!"});
                   })
          } else {
                   callback({message: "Não foi possível inserir Contador!!"});
          }
     },

     updateContador: function(token, descricao, contador, callback) {
            descricao = encodeURIComponent(descricao);
            contador.descricao = encodeURIComponent(contador.descricao);
            contador.prefixo = encodeURIComponent(contador.prefixo);
            contador.contador = encodeURIComponent(contador.contador);
            contador.casas = encodeURIComponent(contador.casas);
            if (descricao != "") {
                   rest.put('http://'+host+":"+port+'/notas/contador/descricao/'+descricao,
                   {
                           accessToken: token,
                           data: contador
                   })
                   .on('success', function(data,response){
                           callback(data);
                   })
                   .on('error', function(err, response){
                           callback({message:"Erro ao atualizar Contador!!"});
                   })
            } else {
                    callback({message: "Não foi possível atualizar Contador!!"});
            }
     },

     getContadores:function(token, callback){
         rest.get('http://'+host+":"+port+'/notas/contador/all',{
                          accessToken: token
                    })
                   .on('success', function(data,response){
                          callback(data);
                   })
                   .on('error', function(err,response){
                          callback({message: "Erro ao buscar Contadores!!"});
                   });
     },

     getContadorByPrefixo:function(token, prefixo, callback){
         prefixo = encodeURIComponent(prefixo);
         if (prefixo != "") {
                  rest.get('http://'+host+":"+port+'/notas/contador/prefixo/'+prefixo,{
                       accessToken: token
                  })
                  .on('success', function(data, response){
                       callback(data);
                  })
                  .on('error', function(err, response){
                       callback({message: "Erro ao buscar Contador!!"});
                  });
         } else {
                  callback({message: "Não foi possível buscar Contador!!"});
         }
     },

     deleteContadorByPrefixo: function(token, prefixo, callback){
         prefixo = encodeURIComponent(prefixo);
         if (prefixo != "") {
                  rest.del('http://'+host+":"+port+'/notas/contador/prefixo/'+prefixo,{
                      accessToken: token
                  })
                  .on('success', function(data, response){
                       callback(data);
                  })
                  .on('error', function(err, response){
                       callback({message: "Erro ao excluir Contador!!"});
                  });
         } else {
            callback({message: "Não foi possível excluir Contador!!"});
         }
     },

     getContadorByDescricao: function(token, descricao, callback){
         descricao = encodeURIComponent(descricao);
         if (descricao != "") {
                 rest.get('http://'+host+":"+port+'/notas/contador/descricao/'+descricao,{
                      accessToken: token
                 })
                 .on('success', function(data, response){
                       callback(data);
                 })
                 .on('error', function(err, response){
                       callback({message: "Erro ao buscar Contador!!"});
                 });
         } else {
                 callback({message: "Não foi possível buscar Contador!!"});
         }
     },

     deleteContadorByDescricao: function(token, descricao, callback){
         descricao = encodeURIComponent(descricao);
         if (descricao != "") {
                 rest.del('http://'+host+":"+port+'/notas/contador/descricao/'+descricao,{
                      accessToken: token
                 })
                 .on('success', function(data, response){
                      callback(data);
                 })
                 .on('error', function(err, response){
                      callback({message: "Erro ao excluir Contador!!"});
                 });
         } else {
                  callback({message: "Não foi possível excluir Contador!!"});
         }
     },

     updateValorContadorByPrefixo: function(token, prefixo, valor, callback){
         prefixo = encodeURIComponent(prefixo);
         valor = encodeURIComponent(valor);
         if (prefixo != "" && valor != "") {
                 rest.put('http://'+host+":"+port+'/notas/contador/prefixo/'+prefixo+'/casas/'+valor,{
                      accessToken: token
                 })
                 .on('success', function(data, response){
                      callback(data);
                 })
                 .on('error', function(err, response){
                      callback({message: "Erro ao atualizar o número de casas do Contador!!"});
                 });
         } else {
                 callback({message: "Não foi possível alterar o número de casas do Contador!!"});
         }
     },

     resetContadorByPrefixo: function(token, prefixo, callback){
         prefixo = encodeURIComponent(prefixo);
         if (prefixo != "") {
                 rest.put('http://'+host+":"+port+'/notas/contador/prefixo/'+prefixo+'/reset',{
                      accessToken: token
                 })
                 .on('success', function(data, response){
                      callback(data);
                 })
                 .on('error', function(err, response){
                      callback({message: "Erro ao reinicializar Contador!!"});
                 });
         } else {
                 callback({message: "Não foi possível reinicializar o contador!!"});
         }
     },

     searchNextContadorByPrefixo: function(token, prefixo, callback){
          prefixo = encodeURIComponent(prefixo);
          if (prefixo != "") {
                 rest.get('http://'+host+':'+port+'/notas/contador/prefixo/'+prefixo+'/searchnext',{
                      accessToken: token
                 })
                 .on('success', function(data, response){
                      callback(data);
                 })
                 .on('error', function(err, response){
                      callback({message: "Erro ao buscar próximo valor para o Contador!!"});
                 });
          } else {
                 callback({message: "Não foi possível buscar o próximo valor!!"});
          }
     },

     getNextContadorByPrefixo: function(token, prefixo, callback){
          prefixo = encodeURIComponent(prefixo);
          if (prefixo != "") {
                 rest.get('http://'+host+':'+port+'/notas/contador/prefixo/'+prefixo+'/next',{
                      accessToken: token
                 })
                 .on('success', function(data, response){
                       callback(data);
                 })
                 .on('error', function(err, response){
                       callback({message: "Erro ao buscar próximo valor para o Contador!!"});
                 });
          } else {
                  callback({message: "Não foi possível buscar o próximo valor!!"});
          }
     },

     searchNextContadorByPrefixo: function(token, prefixo, callback){
          prefixo = encodeURIComponent(prefixo);
          if (prefixo != "") {
                 rest.get('http://'+host+':'+port+'/notas/contador/prefixo/'+prefixo+'/searchnext',{
                      accessToken: token
                 })
                 .on('success', function(data, response){
                       callback(data);
                 })
                 .on('error', function(err, response){
                       callback({message: "Erro ao buscar próximo valor para o Contador!!"});
                 });
          } else {
                  callback({message: "Não foi possível buscar o próximo valor!!"});
          }
     },

     getPriorContadorByPrefixo: function(token, prefixo, callback){
          prefixo = encodeURIComponent(prefixo);
          if (prefixo != "") {
                 rest.get('http://'+host+':'+port+'/notas/contador/prefixo/'+prefixo+'/prior',{
                      accessToken: token
                 })
                 .on('success', function(data, response){
                      callback(data);
                 })
                 .on('error', function(err, response){
                      callback({message: "Erro ao buscar próximo valor para o Contador!!"});
                 });
          } else {
                 callback({message: "Não foi possível buscar o próximo valor!!"});
          }
     },

     incContadorByPrefixo: function(token, prefixo, callback){
          prefixo = encodeURIComponent(prefixo);
          if (prefixo != "") {
                 rest.put('http://'+host+':'+port+'/notas/contador/prefixo/'+prefixo+'/inc',{
                       accessToken: token
                 })
                 .on('success', function(data, response){
                       callback(data);
                 })
                 .on('error', function(err, response){
                       callback({message: "Erro ao incrementar o Contador!!"});
                 });
          } else {
                 callback({message: "Não foi possível incrementar o Contador!!"});
          }
     },

     decContadorByPrefixo: function(token, prefixo, callback){
          prefixo = encodeURIComponent(prefixo);
          if (prefixo != "") {
                 rest.put('http://'+host+':'+port+'/notas/contador/prefixo/'+prefixo+'/dec',{
                      accessToken: token
                 })
                 .on('success', function(data, response){
                       callback(data);
                 })
                 .on('error', function(err, response){
                       callback({message: "Erro ao decrementar o Contador!!"});
                 });
          } else {
                 callback({message: "Não foi possível decrementar o Contador!!"});
          }
     }

};
