// perfilrest/perfilrest.js
var rest = require('restler');
var netConfig = require('netconfig');
var host = netConfig.getHost();
var port = netConfig.getPort();

module.exports = {

   newPerfil: function(token, descricao, callback){
      descricao = encodeURIComponent(descricao.descricao);
      if (descricao != ""){
          rest.post('http://'+host+":"+port+'/notas/perfil/new',
            {
              accessToken: token,
              data: {descricao: descricao}
            })
            .on('success', function(data,response){
                callback(data);
            })
            .on('error', function(err, response){
                callback({message:"Erro ao inserir perfil!!"});
            });
      } else {
          callback({message: "Não foi possível inserir Perfil!!"});
      }
   },

   getPerfils: function(token, callback){
      rest.get('http://'+host+":"+port+'/notas/perfil/all',{
           accessToken: token
      })
      .on('success', function(data,response){
            callback(data);
      })
      .on('error', function(err,response){
             callback({message: "Erro ao buscar Perfil"});
      });
   },

   deletePerfilByDescricao: function(token, descricao, callback){
      if (descricao!="") {
          descricao = encodeURIComponent(descricao);
          rest.del('http://'+host+":"+port+'/notas/perfil/'+descricao,{
              accessToken: token
          })
          .on('success', function(data,response){
              callback(data);
          })
          .on('error', function(err, response){
              callback({message: "Erro ao apagar Perfil!!"});
          });
      } else {
          callback({message: "Não foi possível apagar o Perfil!!"});
      }
   },

   updatePerfilById: function(token, id, descricao, callback){
      if (id != "") {
          descricao = encodeURIComponent(descricao);
          rest.put('http://'+host+':'+port+'/notas/perfil/'+id+'/'+descricao,{
                accessToken: token
          })
          .on('success', function(data,response){
                callback(data);
          })
          .on('error', function(err, response){
                callback({message: "Erro ao atualizar o Perfil!!"});
          });
      } else {
          callback({message: "Não foi possível atualizar o Perfil!!"});
      }
   }

};
